File: algorithms/dp/Dpfibbonaci.ml
(* To link with unix, run ocamlc unix.cma % *)

open Hashtbl;;
open List;;
open Unix;;

let n = 40

(* the dp implementation, loop also uses recursion
 * It's actually quite easy to achieve the same effect as
 * a normal loop.
 * *)
let fib_dp_rec (n: int) : int =
  let a : int ref = ref 0 in
  let b : int ref = ref 1 in
  let rec loop i =
        if i > n then a
        else let tmp = !b
        in  b := !a + !b;
            a := tmp;
            loop (i + 1)
  in !(loop 0)


(* dp 2 with for loop *)
let fib_dp_loop (n: int) : int =
  let a : int ref = ref 0 in
  let b : int ref = ref 1 in
  let tmp : int = 0
  in for i = 0 to n do
    let tmp = !b in
    b := !a + !b;
    a := tmp;
  done;
  !b


(* naive recursion impelementation *)
let rec fib_rec (n : int) : int =
  if n < 2 then 1
  else fib_rec (n - 1) + fib_rec (n - 2)


(* recursion with memoization *)
let fib_rec_mem (n : int) : int =
  let cache = Hashtbl.create 20
  in  Hashtbl.add cache 0 1;
      Hashtbl.add cache 1 1;
      let rec fib_rec_mem_ (n : int) : int =
        match Hashtbl.find_opt cache n with
            Some c -> c
          | None ->
              let res = fib_rec_mem_ (n - 1) + fib_rec_mem_ (n - 2);
              in  Hashtbl.add cache n res;
                  res
      in fib_rec_mem_ n


(* iterative version *)
let fib_iter (n: int) : int =
  let rec fib_iter_ (a: int) (b : int) (n : int) : int =
    if n == 0 then b
    else fib_iter_ b (a + b) (n - 1)
  in fib_iter_ 0 1 n


(* cps transformation *)
let fib_cps (n: int) =
  let ret a b = b in
  let rec fib_cps_ a b n k =
    if n == 0 then k a b
    else fib_cps_ a b (n - 1) (fun a' b' -> k b' (a' + b'))
  in fib_cps_ 0 1 n ret


let timer name f =
  let t = Unix.gettimeofday () in
  let res = f n in
  Printf.printf "%s:, value: %d Excution time %f\n" name res (Unix.gettimeofday () -. t)


let main () =
  timer "fib_dp_rec" fib_dp_rec;
  timer "fib_dp_loop" fib_dp_loop;
  timer "fib_cps" fib_cps;
  timer "fib_rec" fib_rec;
  timer "fib_rec_mem" fib_rec_mem;
  timer "fib_iter" fib_iter


let () = let _ = main () in ()

File: algorithms/dp/dp-fibbonaci.cc
#include <iostream>

#include <chrono>
#include <functional>
#include <unordered_map>

// simple fibbonaci with dp.
// The idea of dp is to separate big problem into smaller prblems,
// solve optimal sub problems, combine solutions to solve the
// big problem.
//
// You can solve a dp problem bottom up or top down.
// Usually top-down implies a normal recursion solution, while
// bottom up will require you to construct the smallest problem then
// goes towards the final answer.
//
// By using dp we can work on looking for optimal subproblems, thus
// ignore suboptimal solutions. In this sense, dp can be think of
// as a way to shrink the possible solution space of a problem.

// dp fib is super simple, but it has several implication. it's
// uses buttom up approach. Since we know after we constructed
// the subsolution, solution for smaller problems can be ignored,
// so we can only use two variables to save the state.

// use dp to construct solution button up.
int fib_dp(int n) {
  int a = 0, b = 1, tmp;
  for (int i = 0; i < n; ++i) {
    tmp = b;
    b = a + b;
    a = tmp;
  }
  return b;
}

// use recursion, a top down approach.
// this solution has several problems:
//  1. It's not tail recursive. Because recursion will defer the
//     evaluation, stack space can be used up when input is large.
//  2. it's possible to recompute sub problems.
int fib_rec(int n) {
  if (n < 2) {
    return 1;
  }
  return fib_rec(n - 1) + fib_rec(n - 2);
}

// solve problem 2 for naive recursion.
// withotu recomputation, this version is as fast as iterative approach.
int fib_rec_mem(int n) {
  static std::unordered_map<int, int> cache{{0, 1}, {1, 1}};

  if (cache.find(n) != cache.end()) {
    return cache[n];
  }

  cache[n] = fib_rec_mem(n - 1) + fib_rec_mem(n - 2);
  return cache[n];
}

// of course we have the iterative method that everybody loves.
int fib_iter(int n) {
  static std::function<int(int, int, int)> fib_iter_ = [](int a, int b,
                                                          int n) -> int {
    if (n == 0) {
      return b;
    }
    return fib_iter_(b, a + b, n - 1);
  };

  return fib_iter_(0, 1, n);
}

// and we have cps transform
// it's trickier then most others.
int fib_cps(int n) {

  static auto ret = [](int a, int b) { return b; };
  static std::function<int(int a, int b, int n, std::function<int(int, int)>)>
      fib_cps_ = [](int a, int b, int n, auto k) -> int {
    if (n == 0) {
      return k(a, b);
    }

    return fib_cps_(a, b, n - 1,
                    [&](int a1, int b1) { return k(b1, a1 + b1); });
  };

  return fib_cps_(0, 1, n, ret);
}

#define N 40

#define START                                                                  \
  {                                                                            \
    auto start = std::chrono::system_clock::now();

#define END                                                                    \
  auto end = std::chrono::system_clock::now();                                 \
  auto elapsed =                                                               \
      std::chrono::duration_cast<std::chrono::milliseconds>(end - start);      \
  printf("time: %ld ms\n", static_cast<long int>(elapsed.count()));            \
  }

int main(void) {

  // 0ms
  START;
  printf("fib_dp: %d\n", fib_dp(N));
  END;

  // 839ms
  START;
  printf("fib_rec: %d\n", fib_rec(N));
  END;

  // 0ms
  START;
  printf("fib_rec_mem: %d\n", fib_rec_mem(N));
  END;

  // 0ms
  START;
  printf("fib_iter: %d\n", fib_iter(N));
  END;

  // 0ms
  START;
  printf("fib_cps: %d\n", fib_cps(N));
  END;

  return 0;
}

#undef N
